doc/RFC-MMP.llnl.04.txt

```diff
@@ -0,0 +1,913 @@
+to the previous design.
+
+=== Multi-modifier protection struct (MMP struct) ===
+
+The MMP structure is a piece of data periodically written to stable storage
+that will allow a node that wants to open a pool in read/write mode to make an
+educated guess of whether another node also has the pool currently open.  It
+should contain at least the following data:
+
+struct mmp_struct {
+      uint64_t mmp_magic;
+      uint64_t mmp_pool_guid;
+      uint64_t mmp_open_id;
+      uint64_t mmp_seq;
+      uint32_t mmp_interval;
+      uint32_t mmp_delay;
+      zio_eck_t mmp_eck;
+
+/* Nice-to-have fields: */
+      char		mmp_nodename[64];
+};
+
+* mmp_magic:     Constant value used for sanity checking.
+* mmp_pool_guid: Pool guid.
+* mmp_open_id:   A random 64-bit value generated when opening the pool in
+		 read/write mode. A special constant value (MMP_CLEAN_ID)
+		 indicates the pool has been closed cleanly.
+* mmp_seq:       A sequence number. Starts at 0 when a pool is first opened
+		 and increases by at least 1, every mmp_interval period.
+		 Also increases by at least 1, at each spa_sync.
+* mmp_interval:  Fixed interval at which MMP blocks are being written
+		 (in milliseconds, usually 5000).  They are also written
+		 independently, based on pool change.
+* mmp_delay:	 Delay being experienced (in milliseconds) by CPU scheduling
+		 + I/O latency.
+* mmp_eck:       Embedded checksum of the MMP data.  To work with existing
+		 zio code, this will be stored at the end of the MMP block,
+		 and is not part of the MMP block itself.
+
+Suggested fields:
+
+* mmp_nodename:  Hostname of the node that wrote this MMP block, for admin
+                 notification purposes.
+
+All nodes must know reliably where to read this data, since these on-disk
+structures are the communication medium between nodes.  This is simplest if the
+data is written to a fixed on-disk position.
+
+=== In-memory structure changes ===
+
+ZFS must be able to detect changes in on-disk MMP blocks, and be able to determine
+whether it is still safe to write to the pool.
+
+A new member of struct spa records the point after which it is unsafe to write
+to the pool.
+
+* spa_mmp_lease:  The latest time, in jiffies, that I/O can safely occur.  This is
+                  based on the last time an MMP block was written, and the
+                  value of mmp_interval.
+
+A new member of struct vdev records the checksums of the MMP blocks as they
+will be on disk if no external actor has altered them.  If they do not match
+upon a read, the MMP blocks were changed on disk by a foreign node, or the
+device contents were corrupted.
+
+* vdev_mmp_array: The checksums of MMP blocks last written to the devices
+
+=== Clean pools or devices ===
+
+A clean pool or device is one which appears not to be in active use, within a
+pool, by an MMP-aware implementation of zfs.
+
+A clean device is one in which all MMP blocks meet the following criteria:
```

Comment:
That makes sense.  Yeah, I'd suggest using a feature flag to ensure that older software doesn't write to the pool.